### PR TITLE
Election votes in separate bucket

### LINKED ISSUE #439: MvP Elections extension
**Is your feature request related to a problem? Please describe.**
In #419 we define an elections extension we need for on-chain governance, given hacking this off chain using multi-sig is not viable.

**Describe the solution you'd like**

We define a super minimal elections process with only "text proposals", which provide proof of agreement, but don't trigger any on-chain action. For this, I see three basic components, which we can extend later for more functionality: Electorate, Election Rules, and Proposal.

`Electorate` defines who may vote in an election. This same group will be used in many elections and we should store it for re-use. It can be of the form of `[]multisig.Participant` or similar data, containing addresses and weights for each participant.

`Election Rules` defines how an election is run. A proposal must be voted upon via a pre-defined ruleset (majority of the senate, 60% of all residents of foo city, etc.). This also includes election length. Here are the basic fields to start with now.

* Title
* Electorate (reference via id)
* Election length (timestamp, eg. 2 days, 1 week, ...)
* Required Vote (integer in steps of 0.01%, so 10000 is 100%. should disallow anything below 5001) - note for now, this is percentage of all **eligible** voters, not the ones who voted. Discuss quorum later

`Proposal` can currently be created by anyone and triggers an election (we will have to refine this quite a bit later). eg. we will need a type field, and text is only present for text proposals. Control who can create a proposal, and which election rules can be used.

* Title
* Text (text string of what is being voted on)
* Election Rules (reference via id)
* Starts at (this must be > block timestamp when tx is included)

**Initialization**

For this first round, electorate and election rules are set in the genesis file with Initializers. We do not yet support creating them with Transactions.

**Transactions**

`CreateProposal` will create a new proposal, must have validate reference to election rules

`Vote(proposal_id, decision)` will vote yes/no/abstain for a given proposal. It will first check:
* headerTime > proposal.starts_at
* headerTime < proposal.starts_at + proposal.rules.length
* voter has not voted on this proposal
* voter is included in proposal.rule.electorate
* Note: clarify who is voting, not just MainSigner.. voter should be in the tx

`Tally(proposal_id)` can only be executed after the election period is over and will update the proposal in the store with:
* total votes yes (sum of weights of all the voters)
* total votes no
* total votes abstain
* total weight of electorate
* passed (boolean)

We assume the election rules and electorate are fixed and cannot change during the course of a running election.

Tally stores information on chain for reference (what was the vote), as well as a passed boolean that is based on the threshold from the rules, so we can filter for passed proposals. In a future iteration, this "passed" flag may trigger an automatic action.

Also, we may consider automatic vote tallying at the end of the period, but that is a different story, let's get the basic pieces together here.

Note: there is request for Acceptance Tests on this. Please add to `bcpd` application then (not bnsd) as this is quite experimental and internal format may/will change without warning. (TODO: future issue)

### PR DESCRIPTION
**Pull request by alpe**
One design issue with the initial proposal was that all votes were stored within the proposal object. While this is quick and ok solution for a small electorate it can become more of a problem with bigger groups as de/serialization and queries become expensive and harder to do.

With this new PR the votes archive is moved into its own bucket. Which let us use the underlying DB features for indexes, queries, etc.

Resolves #439

### CODE REVIEW

File: x/gov/bucket.go [hunk 1]
```diff
@@ -127,3 +127,65 @@ func (b *ProposalBucket) Update(db weave.KVStore, id []byte, obj *TextProposal)
...
+func (b *VoteBucket) Build(db weave.KVStore, proposalID []byte, vote Vote) orm.Object {
+	compositeKey := compositeKey(proposalID, vote.Elector.Signature)
+	return orm.NewSimpleObj(compositeKey, &vote)
+}
+
+func compositeKey(proposalID []byte, address weave.Address) []byte {
+	return append(address, proposalID...)
+}
+
+// HasVote checks the bucket if any vote matching elector address and proposal id was stored.
```
Comment (ruseinov) [inline x/gov/bucket.go]: nit: HasVoted

File: x/gov/handler.go [hunk 2]
```diff
@@ -67,17 +70,21 @@ func (h VoteHandler) Check(ctx weave.Context, db weave.KVStore, tx weave.Tx) (we
...
 	}
-	return res, h.propBucket.Update(db, vote.ProposalID, proposal)
+	if err = h.voteBucket.Save(db, h.voteBucket.Build(db, voteMsg.ProposalID, *vote)); err != nil {
+		return res, errors.Wrap(err, "failed to store vote")
+	}
+	return res, h.propBucket.Update(db, voteMsg.ProposalID, proposal)
 }
 
-func (h VoteHandler) validate(ctx weave.Context, db weave.KVStore, tx weave.Tx) (*VoteMsg, *TextProposal, *Elector, error) {
+func (h VoteHandler) validate(ctx weave.Context, db weave.KVStore, tx weave.Tx) (*VoteMsg, *TextProposal, *Vote, error) {
```
Comment (ruseinov) [inline x/gov/handler.go]: non-blocking: life is hard when we return so many arguments, why don't we wrap all the structs in a struct?

**Comment by ruseinov** [comment]
@alpe github is throwing 500s, so a lot of duplicate messages here..

File: x/gov/handler.go (hunk 2 above)
Comment (alpe) [reply x/gov/handler.go]: Agree that this is a smell. 3 + error is my personal limit to tolerate. Both `Check` and `Deliver` methods are super small so in this special case the caller would not benefit much from a new struct type. But good point and I will keep it in my head for the future.

**Pull request merged**